### ISSUE TITLE
Printing Superset of Labels

### ISSUE THREAD

**Issue by val-co** [abstract-machine-learning/silva#4]
Hello, in the description, you mention that SILVA computes a superset of labels related to a perturbation region. Is there a way to output this superset? 

Also, is there a function to convert a sklearn or XGBoost  tree ensemble to the SILVA format? 

Thanks!

**Comment by marco-zanella** [comment]
Hi,
as an optimization, silva stops when at least two different labels are confirmed, as this is enough to prove lack of stability. We don't have a configuration option to disable this behavior, but the relevant code is in [src/abstract_interpreters/forest_hyperrectangle.c](https://github.com/abstract-machine-learning/silva/blob/master/src/abstract_interpreters/forest_hyperrectangle.c), around line 818, where if a leaf contains a counterexample the analysis stops.

I've never tried silva on XGBoost, but if I recall correctly the structure of an XGBoost tree is similar to that of a CART, i.e. constraints of the form `x_i < k` and a distribution of scores in the leaves. If this is the case it should be simple to adapt the code in [github.com/abstract-machine-learning/data-collection/trainers/classifier_mapper.py](https://github.com/abstract-machine-learning/data-collection/blob/master/trainers/classifier_mapper.py) (see `createForest`), which is basically a depth first visit of each tree in the forest "printing" whether node is a split or a leaf, with some heading. Bear in mind that if XGBoost uses a scoring function different from average or softargmax (which I believe to be the case), that function must be implemented in [src/abstract_interpreters/forest_hyperrectangle.c](https://github.com/abstract-machine-learning/silva/blob/master/src/abstract_interpreters/forest_hyperrectangle.c), as well as its abstract transfer function.

Hope this helps,

**Comment by val-co** [comment]
Do you reckon that the entire superset can be computed in tractable time? 

Also, will the following modification to forest_hyperrectange.c work for exporting the superset?

```
while (!priority_queue_is_empty(Qx)) {
    Hyperrectangle x_prime = priority_queue_pop(Qx);
    const DecisionTreeNode N = priority_queue_pop(Qt);
    unsigned int i;
    double k;
    const unsigned int depth = binary_tree_node_get_depth(N);

    /* A leaf was reached */
    if (decision_tree_node_is_leaf(N)) {
        HyperrectangleDecorator h;
        decorator_create(&h, x_prime, N, x);
        list_push(x->children, h);
        decorator_compute_labels(h->labels, h, data);

        /* Collect all encountered labels */
        set_union(data->all_labels, h->labels);

        /* Leaf contains a counterexample */
        if (set_is_disjoint(h->labels, data->status->labels_a)) {
            data->internal_status = UNSTABLE;
            hyperrectangle_midpoint(status->sample_b, x_prime);
            hyperrectangle_copy(status->region, x_prime);
            * Do not break; continue to collect other labels /
        }

        /* Leaf is "robust", does not help analysis */
        else if (set_is_equal(h->labels, data->status->labels_a)) {
            continue;
        }

        /* Leaf may contain a counterexample: adds to refined decorators */
        list_push(refined, h);
        continue;
    }
}
```

**Comment by marco-zanella** [comment]
It has been a while since I worked on silva, I think you have to replace the `break` on line 823 with a `continue`, to skip the piece of code after the `if` body (which handles split nodes).

Regarding time, the worst-case scenario is exponential, which may happen when every single path but one is labeled as A, the other as B, and the latter is visited last. The `compute_priority` function on line 904 uses a heuristic to guide the best-first search, and we usually get good results in practice. Keep in mind that you can set a timeout, when doing so silva will output a sound superset of the labels if the timeout for a sample is hit. You can have false positives here, but never false negatives.

Hope this helps,
